### PR TITLE
Material text field

### PR DESCRIPTION
**Pull request by tomivirkki**
Closes https://github.com/vaadin/vaadin-text-field/issues/46

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-themes/1)

### CODE REVIEW

**Comment by manolo** [comment]
Review status: 0 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[material/material-theme-text-field.html, line 30 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh6WJHiSWQP9QYE_5tL:-Kh6WJHiSWQP9QYE_5tM:b-wurdsj) ([raw file](https://github.com/vaadin/vaadin-themes/blob/09a1a4d0032b185223ec97a467dd116a794e3651/material/material-theme-text-field.html#L30)):*
> ```HTML
>       }
> 
>       [part="input"] {
> ```

Probably you should use `~=` CSS operator instead as is recommended in the document

---

*[test/visual/import-theme.js, line 1 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh6YVlo0sqPjELrYWjW:-Kh6YVlo0sqPjELrYWjX:b-pirsk7) ([raw file](https://github.com/vaadin/vaadin-themes/blob/09a1a4d0032b185223ec97a467dd116a794e3651/test/visual/import-theme.js#L1)):*
> ```JavaScript
> function importHref(href) {
> ```

Isn't already there a function for this `Polymer.importHref` ?

---

*[test/visual/import-theme.js, line 9 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh6WqALNQmruluGL6vb:-Kh6WqALNQmruluGL6vc:bz8636z) ([raw file](https://github.com/vaadin/vaadin-themes/blob/09a1a4d0032b185223ec97a467dd116a794e3651/test/visual/import-theme.js#L9)):*
> ```JavaScript
> 
> (function() {
>   if (this.HTMLImports) {
> ```

Why do you test HTMLImports here ? 

---

*[test/visual/vaadin-text-field.html, line 61 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh6YjuQXKGDdA_QOeJG:-Kh6YjuQXKGDdA_QOeJH:b-9hpzr0) ([raw file](https://github.com/vaadin/vaadin-themes/blob/09a1a4d0032b185223ec97a467dd116a794e3651/test/visual/vaadin-text-field.html#L61)):*
> ```HTML
> 
>   <script>
>     // HTMLImports.whenReady(function() {
> ```

Why not remove this  commented block?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by tomivirkki** [comment]
Review status: 0 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[material/material-theme-text-field.html, line 30 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh6WJHiSWQP9QYE_5tL:-Kh7_QPoYo1aW5Q12hZd:bnu4rih) ([raw file](https://github.com/vaadin/vaadin-themes/blob/09a1a4d0032b185223ec97a467dd116a794e3651/material/material-theme-text-field.html#L30)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Probably you should use `~=` CSS operator instead as is recommended in the document
</blockquote></details>

Hmm...feels kind of risky. If later we decide to expose part="label-and-input-container" for example then that will get targeted also. Is there some benefit in using it generally/here @jouni ?

---

*[test/visual/import-theme.js, line 1 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh6YVlo0sqPjELrYWjW:-Kh7aQnenAorsZdvR8lk:b-oet2mg) ([raw file](https://github.com/vaadin/vaadin-themes/blob/09a1a4d0032b185223ec97a467dd116a794e3651/test/visual/import-theme.js#L1)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Isn't already there a function for this `Polymer.importHref` ?
</blockquote></details>

Polymer isn't imported at this point yet and we must make sure the theme module is imported when `<vaadin-text-field>` upgrades.

Also `Polymer.importHref`is part of the legacy `polymer.html` that we want to avoid.

---

*[test/visual/import-theme.js, line 9 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh6WqALNQmruluGL6vb:-Kh7_wzHrRlhqIZ1pv0J:be2kfv8) ([raw file](https://github.com/vaadin/vaadin-themes/blob/09a1a4d0032b185223ec97a467dd116a794e3651/test/visual/import-theme.js#L9)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Why do you test HTMLImports here ? 
</blockquote></details>

Because of Gemini. It runs every .js file in the test/visual folder, and running this stuff outside browser will cause an error because "window" is undefined. Maybe it could be configured to target `test.js` only but at least adding that to the `rootUrl` didn't seem to do the trick.

---

*[test/visual/vaadin-text-field.html, line 61 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh6YjuQXKGDdA_QOeJG:-Kh7e6LTLNsbyaS2oksa:b-ihlwhf) ([raw file](https://github.com/vaadin/vaadin-themes/blob/09a1a4d0032b185223ec97a467dd116a794e3651/test/visual/vaadin-text-field.html#L61)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Why not remove this  commented block?
</blockquote></details>

Oh, right. Was supposed to hide the caret somehow but forgot to do it. Will fix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by tomivirkki** [comment]
*[test/visual/vaadin-text-field.html, line 61 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh6YjuQXKGDdA_QOeJG:-Kh7q789i5jZVCvKEBnu:b391xh9) ([raw file](https://github.com/vaadin/vaadin-themes/blob/09a1a4d0032b185223ec97a467dd116a794e3651/test/visual/vaadin-text-field.html#L61)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Oh, right. Was supposed to hide the caret somehow but forgot to do it. Will fix.
</blockquote></details>

Can't seem to get a stable build with the auto-focused field. Removed it for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by platosha** [comment]
Review status: 0 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[material/material-theme-text-field.html, line 30 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh6WJHiSWQP9QYE_5tL:-Kh80cTLepG142hvVuq7:b-3dz3qf) ([raw file](https://github.com/vaadin/vaadin-themes/blob/09a1a4d0032b185223ec97a467dd116a794e3651/material/material-theme-text-field.html#L30)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Hmm...feels kind of risky. If later we decide to expose part="label-and-input-container" for example then that will get targeted also. Is there some benefit in using it generally/here @jouni ?
</blockquote></details>

FYI `[part~="input"]` does not target `part="label-and-input-contaner"`. Only if it’s whitespace-separated.

The benefit of using `~=` is that it allows to expose several part names for a single element. Gives us more flexibility in the DOM that way. :thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by platosha** [comment]
Reviewed 12 of 18 files at r1, 2 of 2 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by tomivirkki** [comment]
*[material/material-theme-text-field.html, line 30 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh6WJHiSWQP9QYE_5tL:-KhLVErxAQpWsDEOPMp0:bgowybl) ([raw file](https://github.com/vaadin/vaadin-themes/blob/09a1a4d0032b185223ec97a467dd116a794e3651/material/material-theme-text-field.html#L30)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary><blockquote>

FYI `[part~="input"]` does not target `part="label-and-input-contaner"`. Only if it’s whitespace-separated.

The benefit of using `~=` is that it allows to expose several part names for a single element. Gives us more flexibility in the DOM that way. :thumbsup: 
</blockquote></details>

Ah, true that. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by jouni** [comment]
The specs for the MD text fields have just updated, and I think we should follow the new spec instead: https://material.io/guidelines/components/text-fields.html#text-fields-field-variations

Also, we should at some point support the “dense“ variation. I’m not pushing that in the scope of this PR/issue, but just a heads-up that we should have that in the end as well.

**Comment by jouni** [comment]
Just noticed myself, that the new specs don’t actually touch the default text field styles, so the styles in this PR are still fine.

File: material/material-theme-text-field.html [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        /*
+        -webkit-transition needed for Safari right now, but it’s fixed in
+        Safari Technology Preview, so we can expect it to fix itself in the
+        near future (Safari 10.2 most probably).
+        */
+        -webkit-transition: background-size 0.25s;
+        transition: background-size 0.25s;
+
+        background-size: 0% 2px, 100% 2px;
+        background-image: linear-gradient(#3F51B5, #3F51B5), linear-gradient(rgba(0, 0, 0, 0.12) 1px, transparent 1px);
```
Comment (jouni) [inline material/material-theme-text-field.html]: Underline color is too low contrast according to the specs. Should be 0.42 instead.

File: material/material-theme-text-field.html [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        background-size: 100% 2px;
+      }
+
+      :host([invalid]) [part~="label"] {
+        color: #dd2c00;
+      }
+
+      :host([disabled]) {
+        color: rgba(0, 0, 0, 0.38);
+        background-image: linear-gradient(90deg, rgba(0, 0, 0, 0.24) 0, rgba(0, 0, 0, 0) 2px);
```
Comment (jouni) [inline material/material-theme-text-field.html]: Underline color is too low contrast according to the specs. Should be 0.42 instead.

**Comment by Saulis** [comment]
- I've enabled the build in Travis, hoping to see a green build before this gets merged.
- Maybe a `demo/index.html` would be also nice (can be the material text field at this point for example.)

---

Review status: 17 of 18 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-:-KhMk1NnIrrzZEVbBBBY:b83amd2)*

**Comment by manolo** [comment]
For some reason build was not run, adding some stuff  (README content) to trigger it and check if CI passes

---

Reviewed 1 of 1 files at r5.
Review status: 18 of 19 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-:-KhO6wDOY7UMu4Ro9mRS:b-ywn6px)*

**Comment by tomivirkki** [comment]
Review status: 15 of 19 files reviewed at latest revision, 4 unresolved discussions.

---

*[material/material-theme-text-field.html, line 25 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh1vmGJKixOVzGLY7cX-r4-25:-KhQjosEd6AiAQFPnOub:b-896fix) ([raw file](https://github.com/vaadin/vaadin-themes/blob/b7e38be3ae388babb2ed5060e4405e058a4f9086/material/material-theme-text-field.html#L25)):*
<details><summary><i>Previously, jouni (Jouni Koivuviita) wrote…</i></summary><blockquote>

Underline color is too low contrast according to the specs. Should be 0.42 instead.
</blockquote></details>

Done.

---

*[material/material-theme-text-field.html, line 95 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-Kh1vmGJKixOVzGLY7cX-r4-95:-KhQkMqr5m4HHIJewZv6:b-896fix) ([raw file](https://github.com/vaadin/vaadin-themes/blob/b7e38be3ae388babb2ed5060e4405e058a4f9086/material/material-theme-text-field.html#L95)):*
<details><summary><i>Previously, jouni (Jouni Koivuviita) wrote…</i></summary><blockquote>

Underline color is too low contrast according to the specs. Should be 0.42 instead.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by tomivirkki** [comment]
@jouni As you guessed this is mostly about bootstrapping the repo and relocating the existing theme from `<vaadin-text-field>` repo as described in https://github.com/vaadin/vaadin-text-field/issues/46.

I consider https://github.com/vaadin/vaadin-themable-mixin/issues/1 to be more about polishing the theme (although the issue is currently in a wrong repo).

**Comment by manolo** [comment]
Reviewed 1 of 1 files at r4.
Review status: 15 of 19 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[.travis.yml, line 16 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-KhQn5ukFoQDKAKRWNOh:-KhQn5ukFoQDKAKRWNOi:bk4fuvt) ([raw file](https://github.com/vaadin/vaadin-themes/blob/b535c65fcf040b679110b0f7d050a5ffe730be9d/.travis.yml#L16)):*
> ```YAML
> install:
>   - npm install
>   - npm install -g bower
> ```

add bower as a dev dependency in package.json and run just npm install. This avoids to run twice npm in travis as well as take advantage of bower being cached in the node_modules cache

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by manolo** [comment]
Reviewed 8 of 19 files at r1, 1 of 2 files at r2, 4 of 4 files at r3, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by tomivirkki** [comment]
*[.travis.yml, line 16 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-KhQn5ukFoQDKAKRWNOh:-KhQnsVeR5L23E2azWtL:byl7daf) ([raw file](https://github.com/vaadin/vaadin-themes/blob/b535c65fcf040b679110b0f7d050a5ffe730be9d/.travis.yml#L16)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

add bower as a dev dependency in package.json and run just npm install. This avoids to run twice npm in travis as well as take advantage of bower being cached in the node_modules cache
</blockquote></details>

Not a concern of this PR. This kind of fine tuning is something that can be done in the skeleton first and then applied directly to all other repos.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by manolo** [comment]
Review status: 16 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[.travis.yml, line 16 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-KhQn5ukFoQDKAKRWNOh:-KhQtd3rMZaomhUZphbV:b-wqjkn3) ([raw file](https://github.com/vaadin/vaadin-themes/blob/b535c65fcf040b679110b0f7d050a5ffe730be9d/.travis.yml#L16)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Not a concern of this PR. This kind of fine tuning is something that can be done in the skeleton first and then applied directly to all other repos.
</blockquote></details>

Oh I see, it was already done in skeleton 1.x branch and other elements but not ported to the skeleton.
So not a blocker if you want to leave as it is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by tomivirkki** [comment]
*[.travis.yml, line 16 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-KhQn5ukFoQDKAKRWNOh:-KhQuWF5NSzNjydADx-C:b-iir8nc) ([raw file](https://github.com/vaadin/vaadin-themes/blob/b535c65fcf040b679110b0f7d050a5ffe730be9d/.travis.yml#L16)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Oh I see, it was already done in skeleton 1.x branch and other elements but not ported to the skeleton.
So not a blocker if you want to leave as it is.
</blockquote></details>

Didn't know it was fixed in the old skeleton..wonder why it didn't make it to the new one. Anyway, I'd rather do it the standard way (trough skeleton) so we don't get 1 repo that's different from others. Can you make the PR to the skeleton?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by manolo** [comment]
Reviewed 1 of 19 files at r1, 1 of 2 files at r2, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[.travis.yml, line 16 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-KhQn5ukFoQDKAKRWNOh:-KhQvGBwzNi0-XCeBVdY:bym16lh) ([raw file](https://github.com/vaadin/vaadin-themes/blob/b535c65fcf040b679110b0f7d050a5ffe730be9d/.travis.yml#L16)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Didn't know it was fixed in the old skeleton..wonder why it didn't make it to the new one. Anyway, I'd rather do it the standard way (trough skeleton) so we don't get 1 repo that's different from others. Can you make the PR to the skeleton?
</blockquote></details>

Actually it's not an old skeleton, but a different branch still used for hybrid elements. Yes I'll do the PRs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1)*

**Comment by manolo** [comment]
:lgtm: 

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-:-KhQvdsOJKH3l0nuLpIi:bg7igkv)*

**Comment by Saulis** [comment]
:lgtm:

---

Reviewed 9 of 19 files at r1, 2 of 2 files at r2, 1 of 4 files at r3, 5 of 5 files at r6, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-:-KhQwV_RNsc6XGkqgYs9:bnfp4nl)*

**Comment by platosha** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-:-KhR1o9pYuj0r88u_XGY:bnfp4nl)*

**Comment by jouni** [comment]
@tomivirkki 
> I consider vaadin/vaadin-themable-mixin#1 to be more about polishing the theme (although the issue is currently in a wrong repo).

If that’s the case, this LGTM, and I can work on the polished CSS in another branch, and try to fix the Safari regressions.

Feel free to merge this.

**Comment by jouni** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-themes/1#-:-KhR_OaOEt2vWBdp_U09:bnfp4nl)*

**Pull request merged**